proposals/0000-custom-type-errors.rst

```diff
@@ -0,0 +1,188 @@
+
+  * X does not implement equality.
+    Maybe you want to derive it using `deriving Eq`?
+  * In the expression: ...
+
+We definitely could include a generic framework to hint auto-derivation of built-in classes such as ``Eq`` or ``Ord``. But in the current Haskell ecosystem some libraries use generic deriving, others Template Haskell, and so on, so a generic facility seems useful.
+
+Inequalities
+~~~~~~~~~~~~
+The ability to attach a custom error message whenever an equality constraint fails is helpful in a variety of scenarios. The simpler one is simply explaining that two arguments should coincide. For example, if ``(==)`` is applied to terms of different type, we could say ::
+
+  * == is applied to arguments of different types `Bool` and `Int`
+  * In the expression: ...
+
+Another possibility is to give a description of the involved types in domain-specific terms. Take for example the `diagrams library <https://hackage.haskell.org/package/diagrams-core-1.4/docs/Diagrams-Core.html#t:QDiagram>`_, it defines a ``QDiagram b v n m`` type indexed by back-end, vector space, number field and annotation types. If we try to combine two diagrams of different vector space, we get a message similar to ::
+
+  * Couldn't match expected type 'V2' with actual type 'V3'
+
+Instead, we could express this message in a more meaningful way ::
+
+  * Couldn't match vector space 'V2' with 'V3'
+
+
+Proposed Change Specification
+-----------------------------
+This proposal defines two new "constraint combinators" to express constraint behavior in the event of a type error.
+
+The first combinator is ``IfUndischarged``. In short, ``IfUndischarged c msg`` shows the message when the constraint ``c`` is left undischarged by the constraint solver. A simple example using ``Eq`` ::
+
+    (==) :: IfUndischarged (Eq a) (ShowType a :<>: Text " does not implement equality."
+                                  :$$: Text "Maybe you want to derive it using `deriving Eq`?")
+         => a -> a -> Bool
+
+If it turns out that ``(==)`` is applied to a type for which no ``Eq`` instance exists, the error message becomes ::
+
+  * X does not implement equality.
+    Maybe you want to derive it using `deriving Eq`?
+  * In the expression: ...
+
+The second combinator is ``IfApartUnsafe`` (the naming shall become clear later). In short, ``IfApartUnsafe a b no yes`` rewrites to ``no`` if ``a`` and ``b`` are apart *at that point in the solving process* and to ``yes`` otherwise. Note that this is *extremely unsafe*, since any knowledge about ``a`` and ``b`` is lost if the second branch is taken, and for that reason we should only expose the variant ::
+
+    type IfApart a b no yes = IfApartUnsafe a b no (a ~ b, yes)
```

Comment:
While this is better for type inference than `IfApartUnsafe`, it still can exhibit unpredictable behavior. Suppose we have `IfApart a b (TypeError (Text "apart")) ()`. Then, sometimes we can get the error message `apart` (if the types are apart at the time when the solver sees the `IfApartUnsafe`) and sometimes we'll get the standard `Cannot unify a with b` (otherwise). This unpredictability could be confusing for users.